### PR TITLE
Refactored Form Handling management

### PR DESCRIPTION
**Pull request by mickaelandrieu**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | As pointing by both @alegout and @sarjon, form handling can be refactored: this is my proposal, see docs => https://github.com/PrestaShop/prestafony-project/blob/fbbd8cb0e77fd94deb6871699cdee74d9652c5f0/docs/migration-guide.md#form-data-handlers
| Type?         | improvement
| Category?     | BO
| BC breaks?    | no
| Deprecations? | yes `PerformanceFormHandler`
| Fixed ticket? | none, but can avoid errors that come from copy/pastes: i've already spotted a lot in current form handlers (repetitive tasks aren't funny btw)
| How to test?  | All migrated pages from Advanced and Shop parameters should works as expected.

## TLDR;

How to register a new form handler?

```yaml
    prestashop.adapter.administration.form_handler:
        class: 'PrestaShop\PrestaShop\Core\Form\FormHandler'
        arguments:
            - '@=service("form.factory").createBuilder()'
            - '@prestashop.hook.dispatcher'
            - '@prestashop.adapter.administration.form_provider'
            - # form Types
              'general': 'PrestaShopBundle\Form\Admin\AdvancedParameters\Administration\GeneralType'
              'upload_quota': 'PrestaShopBundle\Form\Admin\AdvancedParameters\Administration\UploadQuotaType'
              'notifications': 'PrestaShopBundle\Form\Admin\AdvancedParameters\Administration\NotificationsType'
            - 'AdministrationPage' # hookName
```

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8908)

### CODE REVIEW

File: src/PrestaShopBundle/Form/Admin/ShopParameters/General/MaintenanceFormDataProvider.php [hunk 1]
```diff
@@ -37,11 +37,9 @@
     /**
      * @var MaintenanceConfiguration
      */
-    protected $maintenanceConfiguration;
+    private $maintenanceConfiguration;
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Form/Admin/ShopParameters/General/MaintenanceFormDataProvider.php:40]: final class can't have protected properties, or at least it's useless :rofl:

File: src/PrestaShopBundle/Form/Admin/ShopParameters/General/MaintenanceFormHandler.php [hunk 2]
```diff
@@ -1,83 +0,0 @@
-<?php
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Form/Admin/ShopParameters/General/MaintenanceFormHandler.php:1]: I remove every form handler created in develop and not available in 1.7.3.x

File: src/PrestaShopBundle/PrestaShopBundle.php [hunk 3]
```diff
@@ -55,7 +54,6 @@ public function build(ContainerBuilder $container)
         $container->addCompilerPass(new DynamicRolePass());
         $container->addCompilerPass(new PopulateTranslationProvidersPass());
         $container->addCompilerPass(new LoadServicesFromModulesPass(), PassConfig::TYPE_BEFORE_REMOVING);
-        $container->addCompilerPass(new PopulateFormHandlersPass(), PassConfig::TYPE_BEFORE_REMOVING);
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/PrestaShopBundle.php:58]: Compiler pass is not needed anymore.

File: src/PrestaShopBundle/Resources/config/services/adapters.yml [hunk 4]
```diff
@@ -128,25 +124,6 @@ services:
             - '@prestashop.adapter.upload_quota.configuration'
             - '@prestashop.adapter.notifications.configuration'
 
-    prestashop.adapter.performance.form_handler:
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Resources/config/services/adapters.yml]: I moved form handlers configuration to his own files.

**Comment by mickaelandrieu** [comment]
Ping @Quetzacoalt91 @alegout  @sarjon: what do you think?

**Comment by sarjon** [comment]
@mickaelandrieu for me the idea was to change form handler interface to hopefully make it more flexible and more aware of context in which it is handled.

From:

```php
/**
 * @return FormInterface
 */
function getForm();

/**
 * @return array $errors
 */
function save(array $data);
```
To:
```php
/**
 * @return FormInterface
 */
function getForm();

/**
 * Process form returned by getForm() methods.
 * Here you can do anything from validation & saving to dispatching events or even calling API's.
 *
 * We could also make form handler be aware of the request (optionally)
 * since it can be useful in some cases.
 *
 * @return array $errors if any or maybe FormInterface (which could have errors if any)?
 */
function processForm(FormInterface $form, Request $request = null);
```

Probably it's not the best solution, but maybe it can give some ideas? :)

**Comment by mickaelandrieu** [comment]
> Probably it's not the best solution, but maybe it can give some ideas? :)

I don't see the gain except we'll move request processing into form handler and we need to introduce a BC for Maintenance and Performance pages to rely on a new interface. 

Can you add an exemple of a form handler using this new interface ?

**Comment by mickaelandrieu** [comment]
> we need to introduce a BC for Maintenance

in fact easy workaround: keep `save` and `processForm` and add a new interface that extend the existing one.

**Comment by sarjon** [comment]
@mickaelandrieu here's a quick example:

```php
function processForm(FormInterface $form, Request $request)
{
        $form->handleRequest($request);

        if ($form->isSubmitted()) {
            $this->validate($form);

            if (!$form->getErrors()) {
                $data = $form->getData();

                $this->cacheClearer->clearSmartyCache();
                $this->cacheClearer->clearMediaCache();

                if (!$data['stock']['stock_management']) {
                    $data['stock']['allow_ordering_oos'] = 1;
                    $data['page']['display_quantities'] = 0;
                }

                $this->formDataProvider->setData($data);
            }
        }

        $this->dispatcher->dispatchForParameters('actionPPreferencesPageProcessAfter', [
            'form' => $form,    // contains all form errors (if any) & form data
        ]);

        return $form;
}
```

I think it's too early to tell whether it's the right way to do it or not, but maybe this approach could give some flexibility? :confused:

**Comment by mickaelandrieu** [comment]
@sarjon this is far too many responsibilities for only one class, try to keep as SOLID as we can, without creating millions of classes.

What do you think @eternoendless?

**Comment by sarjon** [comment]
well, i totally agree, but maybe example could be improved as well.

```php
function processForm(FormInterface $form, Request $request)
{
        $form->handleRequest($request);

        if ($form->isSubmitted()) {
            // maybe have dedicated form validators for larger forms?
            $this->formValidator->validate($form);   

            if (!$form->getErrors()) {
                $data = $form->getData();

                // maybe dispatch event and create dedicated event listener to clear cache after form is saved?
                // in this case cacheClearer dependency could be removed
                $this->cacheClearer->clearSmartyCache();
                $this->cacheClearer->clearMediaCache();

                $this->formDataProvider->setData($data);
            }
        }

        $this->dispatcher->dispatchForParameters('actionPPreferencesPageProcessAfter', [
            'form' => $form,    // contains all form errors (if any) & form data
        ]);

        return $form;
}
```

**Comment by mickaelandrieu** [comment]
@sarjon the main issue with this implementation is the number of responsibilities for only one function.

As far as I can see, `processForm()`:
* transform the Request into form Data;
* check if the form has been submitted;
* validate the form with our business rules;
* clear the caches;
*dispatch the save of data;
* dispatch an event;
* return a form;

this is far too much! In common Symfony applications, the controller has theses following responsibilities:
* transform the Request into form Data;
* check if the form has been submitted;
* dispatch an event;

This is why I think we should keep them in controllers for now, without excluding the idea of refactoring later: we have time to think about it during 1.7.5 development time.

PS: What I don't want *for sure* is our "own kind" of Form, different from what is provided by Form component. We need PrestaShop code to be maintained by any Symfony developers, this is the major reason of this migration (for me).

**Comment by sarjon** [comment]
yeah, it does make sense. Thaks for reply!

**Comment by LittleBigDev** [comment]
This PR needs to be rebased after `develop` so the php 7.1 functional tests come green again.

**Review by Quetzacoalt91 — DISMISSED** [review]

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/ProductPreferencesFormHandler.php [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+     * @var CacheClearer
+     */
+    private $cacheClearer;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function save(array $data)
+    {
+        $errors = $this->formDataProvider->setData($data);
```
Comment (sarjon) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/ProductPreferencesFormHandler.php:48]: Correct me if im wrong, but it seems that $this->formDataProvider->setData($data); is called in parent::save($data); as well.
It could be like this maybe:
```php
$errors = parent::save($data);

if (empty($errors)) {
    // if logic here
}

return $errors;
```

Comment (mickaelandrieu) [reply src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/ProductPreferencesFormHandler.php:48]: I didn't know if you needed to save once "first" or only once, it's a minor refactoring.

If you think it's worth doing it this way, I update.

Edit: doing your update, the hook twice will be dispatched twice.

Comment (sarjon) [reply src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/ProductPreferencesFormHandler.php:48]: Maybe the part below should be moved from handler to form data provider so it could be set before saving it? Save could be performed only once then.

```php
if (!$data['stock']['stock_management']) {
   // set correct data logic
}
```

Comment (mickaelandrieu) [reply src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/ProductPreferencesFormHandler.php:48]: yes, it's planned but not in this pr: I need this to be merged before friday when the update you suggest can be done during freeze period.

As this pull request grow and grow I try to keep the size as little as possible (31 files already impacted)

We will do it anyway :+1:

**Comment by mickaelandrieu** [comment]
ping @fatmaBouchekoua e2e tests are failed but I can't explain why... can you help me on this please? :)

**Comment by fatmaBouchekoua** [comment]
Hello @mickaelandrieu ,
Sometimes ChromeDriver attempt to click at the wrong location (x,y), This bug will be fixed on ChromeDriver 2.38. We are waiting for this release to not have this problem again.
When you have a similar problem, try to restart the job.
I've restarted the job and tests passed ;)
Hope I was helpful!

**Comment by mickaelandrieu** [comment]
@fatmaBouchekoua yes, thank you!

**Review by Quetzacoalt91 — DISMISSED** [review]

**Comment by eternoendless** [comment]
Reviewed 8 of 15 files at r1, 27 of 46 files at r3, 4 of 9 files at r4, 1 of 2 files at r5, 5 of 6 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/Core/Form/FormHandler.php, line 66 at r6](https://beta.reviewable.io/reviews/prestashop/prestashop/8908#-LAn3Dd5-xDLTMbtMFLf:-LAn3Dd5-xDLTMbtMFLg:b-zf7zxv) ([raw file](https://github.com/prestashop/prestashop/blob/b4f0da72d9536549443296da98d290d7f5e251af/src/Core/Form/FormHandler.php#L66)):*
> ```PHP
>         FormDataProviderInterface $formDataProvider,
>         array $formTypes,
>         $hookName
> ```

Shouldn't this be `$formName` or `$pageName` instead?

---

*[src/Core/Form/FormHandler.php, line 87 at r6](https://beta.reviewable.io/reviews/prestashop/prestashop/8908#-LAn2Xu41yTV4PNdldL_:-LAn2Xu41yTV4PNdldLa:buuinal) ([raw file](https://github.com/prestashop/prestashop/blob/b4f0da72d9536549443296da98d290d7f5e251af/src/Core/Form/FormHandler.php#L87)):*
> ```PHP
> 
>         $this->formBuilder->setData($this->formDataProvider->getData());
>         $this->hookDispatcher->dispatchForParameters("display{$this->hookName}Form", [
> ```

I think we have a problem here. `display` hooks are supposed to be used for display content only (ie. modules attached to this hook should return an HTML string). This hook behaves more like an event, so it should be `action`.

---

*[src/Core/Form/FormHandler.php, line 91 at r6](https://beta.reviewable.io/reviews/prestashop/prestashop/8908#-LAn29-o1m_kNzSJSWks:-LAn29-o1m_kNzSJSWkt:b-krgvta) ([raw file](https://github.com/prestashop/prestashop/blob/b4f0da72d9536549443296da98d290d7f5e251af/src/Core/Form/FormHandler.php#L91)):*
> ```PHP
>         ]);
> 
>         return $this->formBuilder->setData($this->formBuilder->getData())->getForm();
> ```

Here you are performing a `->setData()` with `->getData()` on the same object. Why? In case `formBuilder` was modified by the hook?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8908)*

**Comment by mickaelandrieu** [comment]
> Here you are performing a ->setData() with ->getData() on the same object. Why? In case formBuilder was modified by the hook?

Yes.

File: src/Core/Form/FormHandler.php [hunk 6]
```diff
@@ -84,11 +85,16 @@ public function getForm()
         }
 
         $this->formBuilder->setData($this->formDataProvider->getData());
-        $this->hookDispatcher->dispatchForParameters("display{$this->hookName}Form", [
-            'form_builder' => &$this->formBuilder
-        ]);
+        $this->hookDispatcher->dispatchForParameters(
+            "display{$this->hookName}Form",
```
Comment (mickaelandrieu) [inline src/Core/Form/FormHandler.php]: it's also an action hook as this is not about rendering a string content

**Comment by mickaelandrieu** [comment]
> Shouldn't this be $formName or $pageName instead?

it's only used to render the hook name, so... I think it's worth naming it this way

**Comment by mickaelandrieu** [comment]
Be careful, if you change hooks names you need to update the related xml and sql files like https://github.com/PrestaShop/PrestaShop/blob/develop/install-dev/data/xml/hook.xml#L769 and https://github.com/PrestaShop/PrestaShop/blob/develop/install-dev/upgrade/sql/1.7.4.0.sql

**Comment by eternoendless** [comment]
> Here you are performing a ->setData() with ->getData() on the same object. Why? In case formBuilder was modified by the hook?
> 
> Yes.

But why? The `data` property is private and cannot be changed except through `->setData()`.

File: src/Core/Form/FormHandler.php (hunk 6 above)
Comment (eternoendless) [reply src/Core/Form/FormHandler.php]: > it's also an action hook as this is not about rendering a string content

My point exactly

**Comment by mickaelandrieu** [comment]
@eternoendless, yes this step is not required anymore ! well spotted :)

**Comment by mickaelandrieu** [comment]
@eternoendless comments addressed and tests are green!

**Comment by eternoendless** [comment]
:lgtm:

---

Reviewed 3 of 5 files at r7, 3 of 3 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8908#-:-LAsKrf16Gu_llv3Nxeb:bnfp4nl)*

**Review by eternoendless — DISMISSED** [review]

**Comment by mickaelandrieu** [comment]
Thanks for your review @eternoendless!

**Comment by eternoendless** [comment]
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8908)*

**Comment by eternoendless** [comment]
Thank you @mickaelandrieu

**Review by eternoendless — APPROVED** [review]
